stubs/pika/pika/connection.pyi

```diff
@@ -97,41 +103,42 @@ class Parameters:
 
 class ConnectionParameters(Parameters):
     class _DEFAULT: ...
-    blocked_connection_timeout: Incomplete
-    channel_max: Incomplete
+    # The following attributes only exist if set in the constructor.
+    blocked_connection_timeout: float
+    channel_max: int
     client_properties: Incomplete
-    connection_attempts: Incomplete
-    credentials: Incomplete
-    frame_max: Incomplete
-    heartbeat: Incomplete
-    host: Incomplete
-    locale: Incomplete
-    retry_delay: Incomplete
-    socket_timeout: Incomplete
-    stack_timeout: Incomplete
-    ssl_options: Incomplete
-    port: Incomplete
-    virtual_host: Incomplete
-    tcp_options: Incomplete
+    connection_attempts: int
+    credentials: Any
+    frame_max: int
+    heartbeat: int | Callable[[Connection, int], int] | None
+    host: str
+    locale: str
+    retry_delay: int | float
+    socket_timeout: int | float
+    stack_timeout: int | float
+    ssl_options: SSLOptions
+    port: int
+    virtual_host: str
+    tcp_options: Incomplete | None
+
     def __init__(
         self,
-        host=...,
-        port=...,
-        virtual_host=...,
-        credentials=...,
-        channel_max=...,
-        frame_max=...,
-        heartbeat=...,
-        ssl_options=...,
-        connection_attempts=...,
-        retry_delay=...,
-        socket_timeout=...,
-        stack_timeout=...,
-        locale=...,
-        blocked_connection_timeout=...,
-        client_properties=...,
-        tcp_options=...,
-        **kwargs,
+        host: str = ...,
+        port: int = ...,
+        virtual_host: str = ...,
+        credentials: Any = ...,
```

Comment:
The docstring says this should be an instance of `pika.credentials.Credentials`: https://github.com/pika/pika/blob/e1f7c83ad2279ca98d33151e59955262461dd509/pika/connection.py#L571

Reply:
Such a class does not exist, only `PlainCredentials` and `ExternalCredentials` without a common base class. It's also possible to write custom credential classes. I guess ideally we'd use a protocol. I will add a `_Credentials: TypeAlias = Incomplete` type alias for now.